### PR TITLE
Added hardware_arduino package

### PR DESCRIPTION
**Pull request by Suryansh470**

### CODE REVIEW

File: hardware_layer/hardware_arduino/README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+## Overview
+
+This is a ROS package for interacting arduino through the [rosserial_arduino](http://wiki.ros.org/rosserial_arduino) package. The package is meant to actuate thrusters connected to the arduino through a motor driver, and publish the measurements taken from various sensors connected to the arduino onto separate topics.
+
+The `hardware_arduino` package has been tested under [ROS](http://www.ros.org) Kinetic and Ubuntu 16.04 LTS. The source code is released under a [BSD 3-Clause license](LICENSE.md).
+
+The hardware used are as follows:
+* [Arduino MEGA](https://store.arduino.cc/usa/arduino-mega-2560-rev3)
+* [8V-28V, 5Amp Dual DC Motor Driver with Current Sensor](http://www.nex-robotics.com/products/motor-drivers/8v-28v-5amp-dual-dc-motor-driver-with-current-sense.html)
+
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/README.md]: Add depth sensor information here.

File: hardware_layer/hardware_arduino/README.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+## Setting up Arduino
+
+### Dependencies
+
+- [Robot Operating System (ROS)](http://wiki.ros.org) (middleware for robotics),
+- Following ROS Packages: [rosserial](http://wiki.ros.org/rosserial), [rosserial_arduino](http://wiki.ros.org/rosserial_arduino)
+
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/README.md]: Add udev rules info here

File: hardware_layer/hardware_arduino/README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+## Usage
+
+To connect to the arduino, run:
+```
+roslaunch hardware_arduino hardware_nodes.launch
+```
+
+## Nodes
+
+### arduino_node
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/README.md]: Fill this up!

File: hardware_layer/hardware_arduino/README.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+```
+cd ~/catkin_ws
+catkin_make --pkg hardware_arduino
+```
+
+## Usage
+
+To connect to the arduino, run:
+```
+roslaunch hardware_arduino hardware_nodes.launch
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/README.md]: keep launch file name similar to package name. `hardware_nodes` is activating all hardware nodes?

File: hardware_layer/hardware_arduino/firmware/arduino_node.cpp [hunk 5]
```diff
@@ -0,0 +1,183 @@
+#include "ros.h"
+#include <Arduino.h>
+#include <std_msgs/Int32.h>
+#include <std_msgs/Float64.h>
+#include <math.h>
+#include <Wire.h>
+#include "../include/MS5837.h"
+#include "../include/arduino_thrust.h"
+
+#define pwmPinWest 3
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/firmware/arduino_node.cpp]: Put pin configuration in a separate header files, just like we have done in [`fourtran`](https://github.com/AUV-IITK/fourtran)

File: hardware_layer/hardware_arduino/firmware/arduino_node.cpp [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+#define directionPinNorthUp1 24
+#define directionPinNorthUp2 25
+#define directionPinSouthUp1 22
+#define directionPinSouthUp2 23
+
+#define analogPinPressureSensor A0
+
+MS5837 sensor;
+
+arduino_thrust AEAST,AWEST,ANORTHUP,ASOUTHUP,ANORTHSWAY,ASOUTHSWAY;
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/firmware/arduino_node.cpp]: Comment the code.

File: hardware_layer/hardware_arduino/firmware/arduino_node.cpp [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+float last_pressure_sensor_value, pressure_sensor_value;
+std_msgs::Float64 voltage;
+ros::NodeHandle nh;
+
+void TEastCb(const std_msgs::Int32 msg)
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/firmware/arduino_node.cpp]: Put function definitions later. Over here just put the function declaration

File: hardware_layer/hardware_arduino/firmware/arduino_node.cpp [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+#define directionPinSouthUp2 23
+
+#define analogPinPressureSensor A0
+
+MS5837 sensor;
+
+arduino_thrust AEAST,AWEST,ANORTHUP,ASOUTHUP,ANORTHSWAY,ASOUTHSWAY;
+
+float last_pressure_sensor_value, pressure_sensor_value;
+std_msgs::Float64 voltage;
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/firmware/arduino_node.cpp]: depth sensor message should have header message. Use custom messages.

File: hardware_layer/hardware_arduino/firmware/arduino_node.cpp [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+ros::Subscriber<std_msgs::Int32> subPwmNorthSway("/ard/northsway", &TNorthSwayCb);
+ros::Subscriber<std_msgs::Int32> subPwmSouthSway("/ard/southsway", &TSouthSwayCb);
+ros::Subscriber<std_msgs::Int32> subPwmNorthUp("/ard/northup", &TNorthUpCb);
+ros::Subscriber<std_msgs::Int32> subPwmSouthUp("/ard/southup", &TSouthUpCb);
+ros::Publisher ps_voltage("/varun/sensors/pressure_sensor/depth", &voltage);
+
+
+void setup()
+{
+    nh.initNode();
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/firmware/arduino_node.cpp]: Put a check to see if arduino node is connected or not. Just like in `fourtran` repo

File: hardware_layer/hardware_arduino/include/arduino_thrust.h [hunk 10]
```diff
@@ -0,0 +1,25 @@
+#ifndef arduino_thrust_h
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/include/arduino_thrust.h]: The file name doesn't follow Developer's Guide. It should be in `CameCase`

File: hardware_layer/hardware_arduino/include/arduino_thrust.h [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+{
+    public:
+    
+        arduino_thrust();
+    
+        //sets the direction and pwm pins for the thrusters and initiate them as outputs
+        void setPins(int A1, int D1, int D2);
+    
+        //applies analogwrite and digitalwrite functions in arduino
+        void ON(int pwm, bool Dir1, bool Dir2);
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/include/arduino_thrust.h]: DIR1 and DIR2 is calculated using the pwm sign so you don't need to take that as inputs

File: hardware_layer/hardware_arduino/launch/hardware_nodes.launch [hunk 12]
```diff
@@ -0,0 +1,8 @@
+<launch>
+    <!--<node name="yawDirect" pkg="hardware_imu" respawn="true" type="yawDirect"/>-->
+    <!--<node name="serial_node" pkg="hardware_arduino" respawn="true" type="serial_node.sh"/>-->
+    <node name="cameras" output="screen" pkg="hardware_camera" respawn="true" type="vid_pub">
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/launch/hardware_nodes.launch]: Why is camera node being run from here? Doesn't it have its own launch file?

File: hardware_layer/hardware_arduino/scripts/serial_node.sh [hunk 13]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/scripts/serial_node.sh]: File not required. Use rosnode directly. See fourtran.

File: hardware_layer/hardware_arduino/src/Processing.cpp [hunk 14]
```diff
@@ -0,0 +1,98 @@
+#include <ros/ros.h>
```
Comment (Mayankm96) [inline hardware_layer/hardware_arduino/src/Processing.cpp]: This isn't part of hardware_sruino. Instead, put this in a different package named `control_layer`.

**Comment by Mayankm96** [comment]
@Suryansh470 Use this message [underwater_sensor_msgs/html/msg/Pressure](http://docs.ros.org/hydro/api/underwater_sensor_msgs/html/msg/Pressure.html) for pressure sensor.

File: hardware_layer/hardware_arduino/README.md (hunk 1 above)
Comment (Suryansh470) [reply hardware_layer/hardware_arduino/README.md]: added in the next commit

File: hardware_layer/hardware_arduino/README.md (hunk 2 above)
Comment (Suryansh470) [reply hardware_layer/hardware_arduino/README.md]: added in the next commit

File: hardware_layer/hardware_arduino/README.md (hunk 3 above)
Comment (Suryansh470) [reply hardware_layer/hardware_arduino/README.md]: Added in the next commit

File: hardware_layer/hardware_arduino/README.md (hunk 4 above)
Comment (Suryansh470) [reply hardware_layer/hardware_arduino/README.md]: No , modified it and renamed it to the package name ....... added in the next commit

File: hardware_layer/hardware_arduino/scripts/serial_node.sh (hunk 13 above)
Comment (Suryansh470) [reply hardware_layer/hardware_arduino/scripts/serial_node.sh]: Removed the file and made as per fourtran

File: hardware_layer/hardware_arduino/launch/hardware_nodes.launch (hunk 12 above)
Comment (Suryansh470) [reply hardware_layer/hardware_arduino/launch/hardware_nodes.launch]: Removed the camera node code , and modified it .....

File: hardware_layer/hardware_arduino/firmware/arduino_node.cpp (hunk 7 above)
Comment (Suryansh470) [reply hardware_layer/hardware_arduino/firmware/arduino_node.cpp]: Done

File: hardware_layer/hardware_arduino/firmware/arduino_node.cpp (hunk 9 above)
Comment (Suryansh470) [reply hardware_layer/hardware_arduino/firmware/arduino_node.cpp]: Done

File: hardware_layer/hardware_arduino/include/arduino_thrust.h (hunk 10 above)
Comment (Suryansh470) [reply hardware_layer/hardware_arduino/include/arduino_thrust.h]: Done

File: hardware_layer/hardware_arduino/include/arduino_thrust.h (hunk 11 above)
Comment (Suryansh470) [reply hardware_layer/hardware_arduino/include/arduino_thrust.h]: Done

**Comment by Suryansh470** [comment]
All comments have been added and the code is compiling perfectly ! :D

**Comment by Mayankm96** [comment]
I have added the TravisCI right now which checks for the build. Please send the PR again.

**Pull request closed without merging**